{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb2e43a4",
   "metadata": {},
   "source": [
    "# Seance 1 : Introduction to protein alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7338f058",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f0ae26d",
   "metadata": {},
   "source": [
    "## 1.1 Introduction Needleman-Wunsch algorithm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "16057698",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Algorithme qui renvoie l'algorothme de Needleman-Wunsch pour l'alignement global de deux sequences d'ADN\n",
    "def matrix_needleman_wunsch(seq1, seq2, identity_score, substitution_score, gap_penalty):\n",
    "    \"\"\"Implémente l'algorithme de Needleman-Wunsch pour l'alignement global de deux séquences\"\"\"\n",
    "    # Initialisation de la matrice\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    score_matrix = np.zeros((n+1, m+1), dtype=int)\n",
    "    backtrack_matrix = [[[] for _ in range(m+1)] for _ in range(n+1)]\n",
    "\n",
    "    # Remplissage de la première ligne et de la première colonne\n",
    "    for i in range(n):\n",
    "        score_matrix[i][0] = gap_penalty * i\n",
    "    for j in range(m):\n",
    "        score_matrix[0][j] = gap_penalty * j\n",
    "\n",
    "    for i in range(1, n+1):\n",
    "        backtrack_matrix[i][0] = [1]  # UP\n",
    "\n",
    "    for j in range(1, m+1):\n",
    "        backtrack_matrix[0][j] = [2]  # LEFT\n",
    "\n",
    "    # Remplissage de la matrice\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "            if seq1[i-1] == seq2[j-1]:\n",
    "                score = identity_score\n",
    "            else:\n",
    "                score = substitution_score\n",
    "            match = score_matrix[i-1][j-1] + score\n",
    "            delete = score_matrix[i-1][j] + gap_penalty\n",
    "            insert = score_matrix[i][j-1] + gap_penalty\n",
    "\n",
    "            # Mettre tous les maximums dans la matrice de retour en arrière\n",
    "            score_matrix[i][j] = max(match, delete, insert)\n",
    "            if score_matrix[i][j] == match:\n",
    "                backtrack_matrix[i][j].append(0)  # Diagonal\n",
    "            if score_matrix[i][j] == delete:\n",
    "                backtrack_matrix[i][j].append(1)  # Up\n",
    "            if score_matrix[i][j] == insert:\n",
    "                backtrack_matrix[i][j].append(2)  # Left\n",
    "        \n",
    "    return score_matrix, backtrack_matrix\n",
    "\n",
    "def score_alignment(score_matrix):\n",
    "    \"\"\"Retourne le score de l'alignement optimal\"\"\"\n",
    "    return score_matrix[-1][-1]\n",
    "\n",
    "def optimal_alignment_Needleman(seq1, seq2, backtrack_matrix):\n",
    "    \"\"\"Retourne toutes les séquences alignées optimales\"\"\"\n",
    "    i, j = len(seq1), len(seq2)\n",
    "    a1, a2 = \"\", \"\"\n",
    "\n",
    "    while i > 0 or j > 0:\n",
    "        direction = backtrack_matrix[i][j][0]\n",
    "\n",
    "        if direction == 0:  # diagonal\n",
    "            a1 += seq1[i-1]\n",
    "            a2 += seq2[j-1]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif direction == 1:  # up\n",
    "            a1 += seq1[i-1]\n",
    "            a2 += \"-\"\n",
    "            i -= 1\n",
    "        else:  # left\n",
    "            a1 += \"-\"\n",
    "            a2 += seq2[j-1]\n",
    "            j -= 1\n",
    "\n",
    "    return a1[::-1], a2[::-1]\n",
    "\n",
    "    \n",
    "\n",
    "def Needleman_wunsch(seq1, seq2, identity_score, substitution_score, gap_penalty):\n",
    "    \"\"\"Fonction principale pour exécuter l'algorithme de Needleman-Wunsch\"\"\"\n",
    "    score_matrix, backtrack_matrix = matrix_needleman_wunsch(seq1, seq2, identity_score, substitution_score, gap_penalty)\n",
    "    score = score_alignment(score_matrix)\n",
    "    aligned_seq1, aligned_seq2 = optimal_alignment_Needleman(seq1, seq2, backtrack_matrix)\n",
    "    \n",
    "    return score, aligned_seq1, aligned_seq2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93b97d55",
   "metadata": {},
   "source": [
    "## Test 1 Needleman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "dd8d8981",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test de l'algorithme avec des séquences d'exemple\n",
    "seq1 = \"CHAT\"\n",
    "seq2 = \"CAT\"\n",
    "\n",
    "identity_score_test = 1\n",
    "substitution_score_test = -1\n",
    "gap_penalty_test = -2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89ff5a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_test, aligned_seq1_test, aligned_seq2_test = Needleman_wunsch(seq1, seq2, identity_score_test, substitution_score_test, gap_penalty_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3d84bc3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score de l'alignement : 1\n",
      "Séquence 1 alignée : CHAT\n",
      "Séquence 2 alignée : C-AT\n"
     ]
    }
   ],
   "source": [
    "print(\"Score de l'alignement :\", score_test)\n",
    "print(\"Séquence 1 alignée :\", aligned_seq1_test)\n",
    "print(\"Séquence 2 alignée :\", aligned_seq2_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b61571c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ac2e0827",
   "metadata": {},
   "source": [
    "## 1.2 Algorithme de Needleman-Wunsch, avec BLOSUM62"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "dba2b227",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chargement de la matrice BLOSUM62 depuis un fichier texte\n",
    "\n",
    "def load_blosum(filename):\n",
    "    \"\"\"\n",
    "    Load a substitution matrix (e.g. BLOSUM) from a file.\n",
    "\n",
    "    Returns:\n",
    "        dict: {(aa1, aa2): score}\n",
    "    \"\"\"\n",
    "    matrix = {}\n",
    "\n",
    "    with open(filename) as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    # Ignore comments\n",
    "    lines = [line.strip() for line in lines if line.strip() and not line.startswith(\"#\")]\n",
    "\n",
    "    # First non-comment line = column headers\n",
    "    headers = lines[0].split()\n",
    "\n",
    "    # Remaining lines = rows\n",
    "    for line in lines[1:]:\n",
    "        parts = line.split()\n",
    "        row_aa = parts[0]\n",
    "        scores = parts[1:]\n",
    "\n",
    "        for col_aa, score in zip(headers, scores):\n",
    "            matrix[(row_aa, col_aa)] = int(score)\n",
    "\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "afdfc20e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "0\n",
      "11\n"
     ]
    }
   ],
   "source": [
    "blosum = load_blosum(\"blosum62.txt\")\n",
    "\n",
    "print(blosum[(\"A\", \"A\")])  # 4\n",
    "print(blosum[(\"A\", \"G\")])  # 0\n",
    "print(blosum[(\"W\", \"W\")])  # 11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "6bcb269f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix_needleman_wunsch_blosum(seq1, seq2, blosum, gap_penalty):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "\n",
    "    score_matrix = np.zeros((n+1, m+1), dtype=int)\n",
    "    backtrack_matrix = [[[] for _ in range(m+1)] for _ in range(n+1)]\n",
    "\n",
    "    # Initialisation\n",
    "    for i in range(n+1):\n",
    "        score_matrix[i][0] = gap_penalty * i\n",
    "        if i > 0:\n",
    "            backtrack_matrix[i][0] = [1]  # UP\n",
    "\n",
    "    for j in range(m+1):\n",
    "        score_matrix[0][j] = gap_penalty * j\n",
    "        if j > 0:\n",
    "            backtrack_matrix[0][j] = [2]  # LEFT\n",
    "\n",
    "    # Remplissage\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "\n",
    "            aa = seq1[i-1]\n",
    "            bb = seq2[j-1]\n",
    "\n",
    "            match = score_matrix[i-1][j-1] + blosum[(aa, bb)]\n",
    "            delete = score_matrix[i-1][j] + gap_penalty\n",
    "            insert = score_matrix[i][j-1] + gap_penalty\n",
    "\n",
    "            score_matrix[i][j] = max(match, delete, insert)\n",
    "\n",
    "            if score_matrix[i][j] == match:\n",
    "                backtrack_matrix[i][j].append(0)\n",
    "            if score_matrix[i][j] == delete:\n",
    "                backtrack_matrix[i][j].append(1)\n",
    "            if score_matrix[i][j] == insert:\n",
    "                backtrack_matrix[i][j].append(2)\n",
    "\n",
    "    return score_matrix, backtrack_matrix\n",
    "\n",
    "def Needleman_wunsch_blosum(seq1, seq2, blosum, gap_penalty):\n",
    "    score_matrix, backtrack_matrix = matrix_needleman_wunsch_blosum(\n",
    "        seq1, seq2, blosum, gap_penalty\n",
    "    )\n",
    "\n",
    "    score = score_alignment(score_matrix)\n",
    "    aligned_seq1, aligned_seq2 = optimal_alignment_Needleman(\n",
    "        seq1, seq2, backtrack_matrix\n",
    "    )\n",
    "\n",
    "    return score, score_matrix, aligned_seq1, aligned_seq2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c6f85ec",
   "metadata": {},
   "source": [
    "## Test 2 Needleman Blossum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "22bae147",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test de l'algorithme avec des séquences d'exemple\n",
    "seq1 = \"CAT\"\n",
    "seq2 = \"HER\"\n",
    "\n",
    "gap_penalty_test = -8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e333454e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score de l'alignement : -5\n",
      "Séquence 1 alignée : CAT\n",
      "Séquence 2 alignée : HER\n",
      "Matrice [[  0  -8 -16 -24]\n",
      " [ -8  -3 -11 -19]\n",
      " [-16 -10  -4 -12]\n",
      " [-24 -18 -11  -5]]\n"
     ]
    }
   ],
   "source": [
    "score_test, score_matrix_test, aligned_seq1_test, aligned_seq2_test = Needleman_wunsch_blosum(seq1, seq2, blosum, gap_penalty_test)\n",
    "print(\"Score de l'alignement :\", score_test)\n",
    "print(\"Séquence 1 alignée :\", aligned_seq1_test)\n",
    "print(\"Séquence 2 alignée :\", aligned_seq2_test)\n",
    "print(\"Matrice\", score_matrix_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670db8c8",
   "metadata": {},
   "source": [
    "# Seance 2: Alignement de N séquences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d16e07",
   "metadata": {},
   "source": [
    "## Chargement des séquences à aligner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "390db9eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    }
   ],
   "source": [
    "records = list(SeqIO.parse(\"orthologs.fasta\", \"fasta\"))\n",
    "print(len(records))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bad90d8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID : goshawk\n",
      "Name : goshawk\n",
      "Description : goshawk\n",
      "Sequence : MVLSANDKTNVKNVFTKIGGHAEEYGAETLERMFTTYPPTKTYFPHFDLHHGSAQIKAHGKKVVGALIEAVNHIDDIAGALSKLSDLHAQKLRVDPVNFKLLGQCFLVVVAIHHPSVLTPEVHASLDKFLCAVGNVLTAKYR\n",
      "\n",
      "ID : vulture\n",
      "Name : vulture\n",
      "Description : vulture\n",
      "Sequence : MVLSANDKTNVKTVFTKITGHAEDYGAETLERMFITYPPTKTYFPHFDLHHGSAQIKAHGKKVVGALIEAVNHIDDIAGALSKLSDLHAQKLRVDPVNFKLLGQCFLVVVAIHHPSVLTPEVHASLDKFLCAVGNVLTAKYR\n",
      "\n",
      "ID : lesser\n",
      "Name : lesser\n",
      "Description : lesser panda\n",
      "Sequence : MVLSPADKTNVKSTWDKLGGHAGEYGGEALERTFASFPTTKTYFPHFDLSPGSAQVKAHGKKVADALTLAVGHLDDLPGALSALSDLHAHKLRVDPVNFKLLSHCLLVTLACHHPAEFTPAVHASLDKFFSAVSTVLTSKYR\n",
      "\n",
      "ID : giant\n",
      "Name : giant\n",
      "Description : giant panda\n",
      "Sequence : MVLSPADKTNVKATWDKIGGHAGEYGGEALERTFASFPTTKTYFPHFDLSPGSAQVKAHGKKVADALTTAVGHLDDLPGALSALSDLHAHKLRVDPVNFKLLSHCLLVTLASHHPAEFTPAVHASLDKFFSAVSTVLTSKYR\n",
      "\n",
      "ID : moose\n",
      "Name : moose\n",
      "Description : moose\n",
      "Sequence : MVLSATDKSNVKAAWGKVGGNAPAYGAEALERMFLSFPTTKTYFPHFDLSHGSAQVKAHGEKVANALTKAVGHLDDLPGTLSDLSDLHAHKLRVDPVNFKLLSHTLLVTLAAHLPSDFTPAVHASLDKFLANVSTVLTSKYR\n",
      "\n",
      "ID : alligator\n",
      "Name : alligator\n",
      "Description : alligator\n",
      "Sequence : MVLSMEDKSNVKAIWGKASGHLEEYGAEALERMFCAYPQTKIYFPHFDMSHNSAQIRAHGKKVFSALHEAVNHIDDLPGALCRLSELHAHSLRVDPVNFKFLAHCVLVVFAIHHPSALSPEIHASLDKFLCAVSAVLTSKYR\n",
      "\n",
      "ID : axolotl\n",
      "Name : axolotl\n",
      "Description : axolotl\n",
      "Sequence : MFKLSGEDKANVKAVWDHVKGHEDAFGHEALGRMFTGIEQTHTYFPDKDLNEGSFALHSHGKKVMGALSNAVAHIDDLEATLVKLSDKHAHDLMVDPAEFPRLAEDILVVLGFHLPAKFTYAVQCSIDKFLHVTMRLCISKYR\n",
      "\n",
      "ID : duck\n",
      "Name : duck\n",
      "Description : duck\n",
      "Sequence : MVLSAADKTNVKGVFSKIGGHAEEYGAETLERMFIAYPQTKTYFPHFDLSHGSAQIKAHGKKVAAALVEAVNHVDDIAGALSKLSDLHAQKLRVDPVNFKFLGHCFLVVVAIHHPAALTPEVHASLDKFMCAVGAVLTAKYR\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for record in records:\n",
    "    print(\"ID :\", record.id)\n",
    "    print(\"Name :\", record.name)\n",
    "    print(\"Description :\", record.description)\n",
    "    print(\"Sequence :\", str(record.seq))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e90189",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Cluster import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50b7c26d",
   "metadata": {},
   "source": [
    "./bali_score.py RV11.aligned/BB11001.fasta ~/uniprot.fasta\n",
    "python3 bali_score.py RV11.aligned/BB11001.fasta uniprot.fasta\n",
    "Pour comparer les séquences d'alignement avec le fichier téléchargé de uniprot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "912b588a",
   "metadata": {},
   "source": [
    "## Needleman Wunsch par bloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7167eeb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix_needleman_profile_alignment(seqs, seqs_to_align, gap_penalty):\n",
    "    \"\"\"Return matrix of alignment between the two blocks\"\"\"\n",
    "    n = len(seqs[0])\n",
    "    m = len(seqs_to_align[0])\n",
    "    M = np.zeros((n+1,m+1), dtype=int)\n",
    "    profile_score_matrix = np.zeros((n+1,m+1), dtype=int)\n",
    "    backtrack_matrix = [[[] for _ in range(m+1)] for _ in range(n+1)]\n",
    "\n",
    "    #Initialisation\n",
    "    for i in range(1,n+1):\n",
    "        M[i,0] = i*gap_penalty\n",
    "        backtrack_matrix[i][0] = [1]  # UP\n",
    "    for j in range(1,m+1):\n",
    "        M[0,j] = j*gap_penalty\n",
    "        backtrack_matrix[0][j] = [2]  # LEFT\n",
    "\n",
    "    # Pour chaque place dans la matrice\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1,m+1):\n",
    "            c1 = []\n",
    "            c2 = []\n",
    "\n",
    "            #Remplissage des colonnes\n",
    "            for block1 in seqs:\n",
    "                c1.append(block1[i-1])\n",
    "            for block2 in seqs_to_align:\n",
    "                c2.append(block2[j-1])\n",
    "            \n",
    "            profile_score_matrix[i,j] = column_alignment(c1, c2, gap_penalty)\n",
    "\n",
    "            # Calcul M[i,j] = max(diag, up, left)\n",
    "            diag = M[i-1,j-1] + profile_score_matrix[i,j]\n",
    "            up   = M[i-1,j] + gap_penalty\n",
    "            left = M[i,j-1] + gap_penalty\n",
    "\n",
    "            M[i,j] = max(diag, up, left)\n",
    "\n",
    "            if M[i,j] == diag:\n",
    "                backtrack_matrix[i][j].append(0)  # Diagonal\n",
    "            if M[i,j] == up:\n",
    "                backtrack_matrix[i][j].append(1)  # Up\n",
    "            if M[i,j] == left:\n",
    "                backtrack_matrix[i][j].append(2)  # Left\n",
    "    \n",
    "    return M, backtrack_matrix\n",
    "\n",
    "def backtrack_profile(seqs, seqs_to_align, M, backtrack_matrix):\n",
    "    \"\"\"Retourne les deux blocks de séquences alignés selon la matrice de backtrack\"\"\"\n",
    "    n = len(seqs[0])\n",
    "    m = len(seqs_to_align[0])\n",
    "\n",
    "    aseqs = ['' for _ in seqs]\n",
    "    aseqs_to_align = ['' for _ in seqs_to_align]\n",
    "\n",
    "    i, j = n, m\n",
    "    while i > 0 or j > 0:\n",
    "        direction = backtrack_matrix[i][j][0]\n",
    "\n",
    "        if direction == 0:  # Diagonal\n",
    "            for idx, seq in enumerate(seqs):\n",
    "                aseqs[idx] = seq[i-1] + aseqs[idx]\n",
    "            for idx, seq in enumerate(seqs_to_align):\n",
    "                aseqs_to_align[idx] = seq[j-1] + aseqs_to_align[idx]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif direction == 1:  # Up\n",
    "            for idx, seq in enumerate(seqs):\n",
    "                aseqs[idx] = seq[i-1] + aseqs[idx]\n",
    "            for idx in range(len(seqs_to_align)):\n",
    "                aseqs_to_align[idx] = '-' + aseqs_to_align[idx]\n",
    "            i -= 1\n",
    "        else:  # Left\n",
    "            for idx in range(len(seqs)):\n",
    "                aseqs[idx] = '-' + aseqs[idx]\n",
    "            for idx, seq in enumerate(seqs_to_align):\n",
    "                aseqs_to_align[idx] = seq[j-1] + aseqs_to_align[idx]\n",
    "            j -= 1\n",
    "\n",
    "    return aseqs, aseqs_to_align\n",
    "\n",
    "def column_alignment(c1, c2, gap_penalty):\n",
    "    \"\"\"Compare une colonne des deux blocks et renvoie la moyenne du score calculé nxm\"\"\"\n",
    "    n = len(c1)\n",
    "    m = len(c2)\n",
    "    M = np.zeros((n,m))\n",
    "\n",
    "    for i in range(n):\n",
    "        for j in range(m):\n",
    "            a = c1[i]\n",
    "            b = c2[j]\n",
    "            #Si il y a un gap\n",
    "            if a == '-' or b == '-':\n",
    "                M[i,j] = gap_penalty\n",
    "            else:\n",
    "                #Si pas de gap\n",
    "                M[i,j] = blosum[(a, b)]\n",
    "    \n",
    "    return np.sum(M)/(n*m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "49f3f3e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligned seq1:\n",
      "-HER\n",
      "\n",
      "Aligned seq2:\n",
      "C-AT\n",
      "CHAT\n",
      "\n",
      "Score d'alignement: -10\n",
      "[[  0  -8 -16 -24 -32]\n",
      " [ -8  -3  -8 -16 -24]\n",
      " [-16 -11  -7  -9 -17]\n",
      " [-24 -19 -15  -8 -10]]\n"
     ]
    }
   ],
   "source": [
    "seq1 = [\"C-AT\", \"CHAT\"]\n",
    "seq2 = [\"HER\"]\n",
    "gap_penalty_test = -8\n",
    "\n",
    "# Remplissage de la matrice de score et de backtrack\n",
    "M, backtrack_matrix = matrix_needleman_profile_alignment(seq2, seq1, gap_penalty_test)\n",
    "\n",
    "# Backtrack pour obtenir les alignements\n",
    "aligned_seq1, aligned_seq2 = backtrack_profile(seq2, seq1, M, backtrack_matrix)\n",
    "\n",
    "# Affichage\n",
    "print(\"Aligned seq1:\")\n",
    "for s in aligned_seq1:\n",
    "    print(s)\n",
    "print(\"\\nAligned seq2:\")\n",
    "for s in aligned_seq2:\n",
    "    print(s)\n",
    "\n",
    "# Score final\n",
    "print(\"\\nScore d'alignement:\", score_alignment(M))\n",
    "\n",
    "#Matrice\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebfa6709",
   "metadata": {},
   "source": [
    "## Matrice de distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "f9f67041",
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix_distance(seq, gap_penalty, blosum):\n",
    "    \"\"\"Renvoie la matrice de distance avec les scores obtenus avec l'algorithme de Needleman Wunsch Blosum\"\"\"\n",
    "    n = len(seq)\n",
    "    M = np.zeros((n,n), dtype=int)\n",
    "    for i in range(n):\n",
    "        for j in range(i, n):\n",
    "            score_matrix = matrix_needleman_wunsch_blosum(seq[i],seq[j],blosum,gap_penalty)[0]\n",
    "            M[i,j] = score_alignment(score_matrix)\n",
    "            M[j,i] = score_alignment(score_matrix)\n",
    "    \n",
    "    return M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "38b782bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[18 10 -5]\n",
      " [10 26 -2]\n",
      " [-5 -2 18]]\n"
     ]
    }
   ],
   "source": [
    "seq1 = [\"CAT\", \"CHAT\"]\n",
    "seq2 = [\"HER\"]\n",
    "gap_penalty_test = -8\n",
    "\n",
    "print(matrix_distance(seq1 + seq2, gap_penalty_test, blosum))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0401e328",
   "metadata": {},
   "source": [
    "## Algorithme UPGMA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "7129cd5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Node:\n",
    "    def __init__(self, name=None, left=None, right=None, height=0.0, size=1):\n",
    "        \"\"\"\n",
    "        A node in a UPGMA tree\n",
    "        \"\"\"\n",
    "        self.name = name        # label for leaves\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.height = height    # height of the node\n",
    "        self.size = size        # number of leaves under this node\n",
    "\n",
    "    def is_leaf(self):\n",
    "        return self.left is None and self.right is None\n",
    "\n",
    "    def __repr__(self):\n",
    "        if self.is_leaf():\n",
    "            return self.name\n",
    "        return f\"({self.left},{self.right})\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "c1394e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Algorithme UPGMA\n",
    "\n",
    "class UPGMA:\n",
    "    def __init__(self, labels, distance_matrix):\n",
    "        \"\"\"\n",
    "        labels: list of taxon names\n",
    "        distance_matrix: numpy array (NxN)\n",
    "        \"\"\"\n",
    "        self.labels = labels\n",
    "        self.D = distance_matrix.astype(float)\n",
    "\n",
    "        # Initial clusters\n",
    "        self.clusters = {\n",
    "            i: Node(name=labels[i]) for i in range(len(labels))\n",
    "        }\n",
    "\n",
    "    def find_closest_clusters(self):\n",
    "        \"\"\"Return indices (i, j) of closest clusters\"\"\"\n",
    "        min_dist = float(\"inf\")\n",
    "        pair = None\n",
    "\n",
    "        for i in self.clusters:\n",
    "            for j in self.clusters:\n",
    "                if i < j and self.D[i, j] < min_dist:\n",
    "                    min_dist = self.D[i, j]\n",
    "                    pair = (i, j)\n",
    "\n",
    "        return pair, min_dist\n",
    "\n",
    "    def merge_clusters(self, i, j, dist):\n",
    "        ci = self.clusters[i]\n",
    "        cj = self.clusters[j]\n",
    "\n",
    "        new_node = Node(left=ci, right=cj, height=dist/2, size=ci.size + cj.size)\n",
    "\n",
    "        new_index = max(self.clusters) + 1\n",
    "\n",
    "        # Créer une nouvelle matrice plus grande\n",
    "        N = self.D.shape[0]\n",
    "        D_new = np.zeros((N+1, N+1))\n",
    "        D_new[:N, :N] = self.D\n",
    "\n",
    "        # Calcul des distances vers les autres clusters\n",
    "        for k in self.clusters:\n",
    "            if k != i and k != j:\n",
    "                dik = self.D[i, k]\n",
    "                djk = self.D[j, k]\n",
    "                D_new[new_index, k] = (ci.size * dik + cj.size * djk) / (ci.size + cj.size)\n",
    "                D_new[k, new_index] = D_new[new_index, k]\n",
    "\n",
    "        self.D = D_new\n",
    "        self.clusters[new_index] = new_node\n",
    "\n",
    "        # Supprimer les anciens clusters\n",
    "        del self.clusters[i]\n",
    "        del self.clusters[j]\n",
    "\n",
    "        return new_index\n",
    "\n",
    "    def build_tree(self):\n",
    "        \"\"\"Run UPGMA and return the root\"\"\"\n",
    "        while len(self.clusters) > 1:\n",
    "            (i, j), dist = self.find_closest_clusters()\n",
    "            new_idx = self.merge_clusters(i, j, dist)\n",
    "\n",
    "        return next(iter(self.clusters.values()))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0642a0b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = [\"C-AT\", \"CHAT\"]\n",
    "seq2 = [\"HER\"]\n",
    "gap_penalty_test = -8\n",
    "\n",
    "# Remplissage de la matrice de score et de backtrack\n",
    "M, backtrack_matrix = matrix_needleman_profile_alignment(seq2, seq1, gap_penalty_test)\n",
    "\n",
    "# Backtrack pour obtenir les alignements\n",
    "aligned_seq1, aligned_seq2 = backtrack_profile(seq2, seq1, M, backtrack_matrix)\n",
    "\n",
    "# Affichage\n",
    "print(\"Aligned seq1:\")\n",
    "for s in aligned_seq1:\n",
    "    print(s)\n",
    "print(\"\\nAligned seq2:\")\n",
    "for s in aligned_seq2:\n",
    "    print(s)\n",
    "\n",
    "# Score final\n",
    "print(\"\\nScore d'alignement:\", score_alignment(M))\n",
    "\n",
    "#Matrice\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "49cac9f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(CHAT,(CAT,HER))\n"
     ]
    }
   ],
   "source": [
    "seq1 = [\"CAT\", \"CHAT\"]\n",
    "seq2 = [\"HER\"]\n",
    "gap_penalty_test = -8\n",
    "\n",
    "D = matrix_distance(seq1 + seq2, gap_penalty_test, blosum)\n",
    "\n",
    "upgma = UPGMA(seq1 + seq2, D)\n",
    "tree = upgma.build_tree()\n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7e42596",
   "metadata": {},
   "source": [
    "## Tracé de l'arbre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "23b64ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "from Bio.Phylo.Newick import Tree, Clade\n",
    "\n",
    "def node_to_clade(node):\n",
    "    \"\"\"Convert Node UPGMA en Clade Biopython pour affichage\"\"\"\n",
    "    if node.name is not None:  # feuille\n",
    "        return Clade(branch_length=node.height, name=node.name)\n",
    "    else:\n",
    "        left = node_to_clade(node.left)\n",
    "        right = node_to_clade(node.right)\n",
    "        return Clade(branch_length=node.height - node.left.height, clades=[left, right])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "d8235d30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlsAAAGwCAYAAACerqCtAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAspUlEQVR4nO3df1xVVb7/8fcJBFHgmCY/TBS7GhqkGTqJNzOHwnTGBzY2Wk2mNVl21VIzDc3G6TFd6l7HyGvqWKQ55q8JMUcbrzTyQ0u7YmhOKJppGEJem+QoFoju7x9dz7cTiIcfiwOH1/Px2H/stdfa53NWe+a83Xudg82yLEsAAAAw4hpPFwAAAODNCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIF9PF+CNLl26pJMnTyooKEg2m83T5QAAADdYlqWzZ8+qU6dOuuaahrsfRdgy4OTJk4qIiPB0GQAAoA5OnDihzp07N9j5CFsGBAUFSfrhP1ZwcLCHqwEAAO5wOByKiIhwfo43FMKWAZcfHQYHBxO2AABoZhp6CRAL5AEAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAa1iLC1ePFidevWTa1bt1ZsbKx27Nhxxb5ZWVmy2WxVtkOHDjVixQAAwFt4fdhat26dpk6dqjlz5igvL0+DBg3SsGHDVFhYWOO4goICFRcXO7cePXo0UsUAAMCbeH3YWrBggX7729/qscceU69evZSSkqKIiAgtWbKkxnEhISEKCwtzbj4+Po1UMQAA8CZeHbYqKiq0d+9eJSQkuLQnJCToo48+qnFs3759FR4ervj4eGVmZtbYt7y8XA6Hw2UDAACQvDxsnT59WhcvXlRoaKhLe2hoqEpKSqodEx4ermXLliktLU0bNmxQVFSU4uPjlZOTc8XXSU5Olt1ud24REREN+j4AAEDz5evpAhqDzWZz2bcsq0rbZVFRUYqKinLux8XF6cSJE5o/f77uuOOOasckJSVp+vTpzn2Hw0HgAgAAkrz8ztZ1110nHx+fKnexTp06VeVuV00GDBigI0eOXPG4v7+/goODXTYAAADJy8OWn5+fYmNjlZGR4dKekZGhgQMHun2evLw8hYeHN3R5AACgBfD6x4jTp0/X2LFj1a9fP8XFxWnZsmUqLCzUxIkTJf3wCLCoqEgrV66UJKWkpCgyMlLR0dGqqKjQqlWrlJaWprS0NE++DQAA0Ex5fdgaM2aMvvnmG7344osqLi5WTEyM3n//fXXt2lWSVFxc7PKbWxUVFZoxY4aKiooUEBCg6OhobdmyRcOHD/fUWwAAAM2YzbIsy9NFeBuHwyG73a7S0lLWbwEA0EyY+vz26jVbAAAAnkbYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAGgmLMvSsmXLdNtttykwMFDt2rVTv379lJKSovPnz0uS5s2bp1tuuaXK2OPHj8tms2nfvn1VjiUkJMjHx0e7d+926VvTNm/ePIPv1Lv4eroAAADgnrFjx2rDhg16/vnntWjRInXs2FH79+9XSkqKIiMjNXLkyFqfs7CwULt27dLkyZOVmpqqAQMGKCIiQsXFxc4+8+fP19atW/XBBx842wIDAxviLbUIXn9nKycnRyNGjFCnTp1ks9m0cePGq47Jzs5WbGysWrdurRtuuEFLly41XygAADVYv3693nnnHa1Zs0azZ89W//79FRkZqcTERG3fvl1Dhgyp03mXL1+uX/7yl3ryySe1bt06lZWVycfHR2FhYc4tMDBQvr6+VdrgHq8PW2VlZerTp48WLVrkVv9jx45p+PDhGjRokPLy8jR79mw99dRTSktLM1wpAABX9s477ygqKkqJiYlVjtlsNtnt9lqf07IsLV++XA899JB69uypG2+8UevXr2+IcvEjXv8YcdiwYRo2bJjb/ZcuXaouXbooJSVFktSrVy/l5uZq/vz5GjVqVLVjysvLVV5e7tx3OBz1qhkAgJ86cuSIoqKi3Op74MCBKneeLMuq0u+DDz7Q+fPnNXToUEnSQw89pNTUVD3yyCP1LxhOXn9nq7Z27dqlhIQEl7ahQ4cqNzdXFy5cqHZMcnKy7Ha7c4uIiGiMUgEAXqasrMy5AL2srMzlmGVZstlsbp0nKipK+/btc9nef//9Kv1SU1M1ZswY+fr+cO/lgQce0Mcff6yCgoL6vxk4EbZ+oqSkRKGhoS5toaGhqqys1OnTp6sdk5SUpNLSUud24sSJxigVANCC3HjjjTp48KBbff38/NS9e3eXrWvXri59/vnPf2rjxo1avHixfH195evrq+uvv16VlZV66623TLyFFouwVY2f/svh8q3XK/2Lwt/fX8HBwS4bAAAN6cEHH9Thw4f13nvvVTlmWZZKS0trdb533nlHnTt31v79+13ugKWkpOjtt99WZWVlQ5Xe4hG2fiIsLEwlJSUubadOnZKvr686dOjgoaoAAC3d6NGjNWbMGD3wwANKTk5Wbm6uvvzyS23evFl33XWXMjMza3W+1NRU3XfffYqJiXHZHn30UZ05c0Zbtmwx9E5aHsLWT8TFxSkjI8Olbdu2berXr59atWrloaoAAC2dzWbT6tWrtWDBAqWnp2vw4MHq3bu35s2bp8TEROcid3fs3btX+/fvr/aLX0FBQUpISFBqampDlt+i2azqvp7gRc6dO6fPP/9cktS3b18tWLBAQ4YMUfv27dWlSxclJSWpqKhIK1eulPTDTz/ExMToiSee0IQJE7Rr1y5NnDhRa9asueK3EX/K4XDIbrertLSUR4oAALeVlZU5v0V47tw5tW3b1sMVtSymPr+9/qcfcnNzXX7obfr06ZKkcePGacWKFSouLlZhYaHzeLdu3fT+++9r2rRpev3119WpUyctXLjQ7aAFAADwY15/Z8sTuLMFAKgL7mx5lqnPb9ZsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAM8vqwlZOToxEjRqhTp06y2WzauHFjjf2zsrJks9mqbIcOHWqcggEAgFfx9XQBppWVlalPnz565JFHNGrUKLfHFRQUKDg42LnfsWNHE+UBAAAv5/Vha9iwYRo2bFitx4WEhKhdu3YNXxAAAGhRvP4xYl317dtX4eHhio+PV2ZmZo19y8vL5XA4XDYAAACJsFVFeHi4li1bprS0NG3YsEFRUVGKj49XTk7OFcckJyfLbrc7t4iIiEasGAAANGU2y7IsTxfRWGw2m9LT0zVy5MhajRsxYoRsNps2bdpU7fHy8nKVl5c79x0OhyIiIlRaWuqy7gsAgJqUlZUpMDBQknTu3Dm1bdvWwxW1LA6HQ3a7vcE/v7mz5YYBAwboyJEjVzzu7++v4OBglw0AAEAibLklLy9P4eHhni4DAAA0Q17/bcRz587p888/d+4fO3ZM+/btU/v27dWlSxclJSWpqKhIK1eulCSlpKQoMjJS0dHRqqio0KpVq5SWlqa0tDRPvQUAANCMeX3Yys3N1ZAhQ5z706dPlySNGzdOK1asUHFxsQoLC53HKyoqNGPGDBUVFSkgIEDR0dHasmWLhg8f3ui1AwCA5q9FLZBvLKYW2AEAvBsL5D2LBfIAAADNEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADDI19MF4OrKyso8XQIAoBHw//feibDVDAQGBnq6BAAAUEc8RgQAADCIO1vNwLlz5zxdAgCgEZSVlSk0NNTTZaCBEbaagbZt23q6BAAAUEc8RgQAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAoCb7txH37Nmjv/zlLyosLFRFRYXLsQ0bNnioKgAAgNppkne21q5dq3/9139Vfn6+0tPTdeHCBeXn52v79u2y2+2eLg8AAMBtTTJs/fu//7teffVVbd68WX5+fnrttdd08OBBjR49Wl26dPF0eQAAAG5rkmHr6NGj+sUvfiFJ8vf3V1lZmWw2m6ZNm6Zly5Z5uDoAAAD3Ncmw1b59e509e1aSdP311+sf//iHJOnMmTM6f/68J0sDAAColSa5QH7QoEHKyMjQzTffrNGjR+vpp5/W9u3blZGRofj4eE+XBwAA4LYmGbYWLVqk77//XpKUlJSkVq1aaefOnfrVr36luXPnerg6AAAA99X7MaJlWVq2bJluu+02BQYGql27durXr59SUlKqPPL76quv5Ofnp549ezrb5s2bJ5vN5rJ16NBB119/vWw2m3x8fDR69Ght2rRJCxYs0LXXXlvfkgEAABpNvcPW2LFjNXXqVCUmJiozM1P79u3T3Llz9d5772nbtm0ufVesWKHRo0fr/Pnz+vDDDyVJM2bMUHFxsXPr3LmzEhMTXdoiIiIkSZWVlUpKSnK7tuTkZPXv319BQUEKCQnRyJEjVVBQcNVx2dnZio2NVevWrXXDDTdo6dKltZgRAACA/89mWZZV18Hr16/XmDFjtHHjRiUmJrocsyxLDofD+btYlmWpe/fuWrx4sTIzM3Xq1Cm99dZbVc4ZGRmpr7/+WsOHD9cbb7yh9u3bS5IOHTqkBx98UKWlpTp69Khb9d1zzz26//771b9/f1VWVmrOnDk6cOCA8vPz1bZt22rHHDt2TDExMZowYYKeeOIJffjhh/q3f/s3rVmzRqNGjXLrdS+/79LSUgUHB7s1BgCAsrIyBQYGSpLOnTt3xc8qmGHq87teYSsxMVEFBQU6dOjQVftu375dv/nNb/TVV1/p0KFDuu2221RcXKygoCCXfpGRkRo7dqwyMzN17NgxrVixQocPH9azzz6r++67T6+//nqVMe763//9X4WEhCg7O1t33HFHtX1mzZqlTZs26eDBg862iRMnav/+/dq1a5dbr0PYAgDUBWHLs0x9ftfrMeKRI0cUFRXlVt/U1FTdf//98vHxUXR0tLp3765169ZV27dDhw7KycnRfffdp3vuuUfTpk3TW2+9pZUrV9Y5aElSaWmpJDnvllVn165dSkhIcGkbOnSocnNzdeHChWrHlJeXy+FwuGwAAABSPcOWZVmy2WxX7XfmzBlt2LBBDz30kLPtoYceqvYx4mWbN2/WmjVrNHDgQLVr105vvPGGTp48Wa9ap0+frttvv10xMTFX7FdSUqLQ0FCXttDQUFVWVur06dPVjklOTpbdbndul9eYAQAA1Cts3XjjjS6P265k9erV+v7773XbbbfJ19dXvr6+mjVrlnbt2qX8/Pwq/devX6/Ro0dr5syZysnJ0aeffip/f3/dfPPNWr9+fZ1qnTx5sj799FOtWbPmqn1/GiAvP2m9UrBMSkpSaWmpcztx4kSdagQAAN6nXmHrwQcf1OHDh/Xee+9VOWZZlvOxXWpqqp555hnt27fPue3fv19Dhgyp9u7WsWPH9PHHH2vGjBmy2WwKCwvT+++/rxdffFGPPvporeucMmWKNm3apMzMTHXu3LnGvmFhYSopKXFpO3XqlHx9fdWhQ4dqx/j7+ys4ONhlAwAAkOoZtkaPHq0xY8bogQceUHJysnJzc/Xll19q8+bNuuuuu5w/BfHJJ5/oscceU0xMjMv2wAMPaOXKlVXWQj3zzDPq06dPldebNGmS9u7d63Z9lmVp8uTJ2rBhg7Zv365u3bpddUxcXJwyMjJc2rZt26Z+/fqpVatWbr82AACAVM+wZbPZtHr1ai1YsEDp6ekaPHiwevfurXnz5ikxMVFDhw5VamqqbrrpJpcfMr1s5MiR+uc//6m//vWvLu2+vlf+YXt3F+RLP4SzVatWafXq1QoKClJJSYlKSkr03XffOfskJSXp4Ycfdu5PnDhRX375paZPn66DBw/qrbfeUmpqqmbMmOH26wIAAFxWr59+MOndd9/V+vXrVVhYqIqKCpdjn3zyiVvnuNIaq+XLl2v8+PGSpPHjx+v48ePKyspyHs/Ozta0adP02WefqVOnTpo1a5YmTpzodu389AMAoC746QfPMvX53ST/NuLChQs1Z84cjRs3Tu+9954eeeQRHT16VHv27NGkSZPcPo87OXLFihVV2gYPHux2oAMAAKhJvf9cjwmLFy/WsmXLtGjRIvn5+WnmzJnKyMjQU0895Vx0DwAA0Bw0ybBVWFiogQMHSpICAgJ09uxZST/8HUZ3froBAACgqWiSYSssLEzffPONJKlr167avXu3pB9+EqKJLjEDAACoVpMMWz//+c+d31D87W9/q2nTpunuu+/WmDFjdO+993q4OgAAAPc1yQXyc+bM0fXXXy/ph59iaN++vXbu3KkRI0Zo2LBhHq4OAADAfU3ypx98fHxUXFyskJAQl/ZvvvlGISEhunjxoocqcw8//QAAqAt++sGzTH1+N8nHiFfKf+fOnVPr1q0buRoAAIC6a1KPEadPny7phx8jfeGFF9SmTRvnsYsXL+rjjz/WLbfc4qHqAAAAaq9Jha28vDxJP9zZOnDggPz8/JzH/Pz81KdPH/5sDgAAaFaaVNjKzMyUJD3yyCN67bXXWO8EAACavSYVti5bvny5p0sAAABoEE1ygTwAAIC3IGwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQV4dtpKTk9W/f38FBQUpJCREI0eOVEFBQY1jsrKyZLPZqmyHDh1qpKoBAIA38fV0ASZlZ2dr0qRJ6t+/vyorKzVnzhwlJCQoPz9fbdu2rXFsQUGBgoODnfsdO3as9euXlZXJx8en1uMAAC1TWVmZp0uAAV4dtrZu3eqyv3z5coWEhGjv3r264447ahwbEhKidu3aufU65eXlKi8vd+47HA5JUqdOnWpXMAAA8Dpe/Rjxp0pLSyVJ7du3v2rfvn37Kjw8XPHx8crMzKyxb3Jysux2u3OLiIhokHoBAC2TZVlXfQKD5sNmWZbl6SIag2VZSkxM1LfffqsdO3ZcsV9BQYFycnIUGxur8vJy/fnPf9bSpUuVlZV1xbth1d3ZioiI0MmTJ10eRQIA4A6Clmc4HA7Z7XaVlpY26Od3iwlbkyZN0pYtW7Rz50517ty5VmNHjBghm82mTZs2udXf1H8sAABgjqnP7xbxGHHKlCnatGmTMjMzax20JGnAgAE6cuSIgcoAAIC38+oF8pZlacqUKUpPT1dWVpa6detWp/Pk5eUpPDy8gasDAAAtgVeHrUmTJmn16tV67733FBQUpJKSEkmS3W5XQECAJCkpKUlFRUVauXKlJCklJUWRkZGKjo5WRUWFVq1apbS0NKWlpXnsfQAAgObLq8PWkiVLJEl33nmnS/vy5cs1fvx4SVJxcbEKCwudxyoqKjRjxgwVFRUpICBA0dHR2rJli4YPH95YZQMAAC/SYhbINyYWyAMA0PywQB4AAKAZImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEH1CltZWVmy2Ww6c+ZMlWORkZFKSUlx2bfZbFW2l19+WZJ0/Phxl3a73a4BAwbor3/9a31KBAAA8CjfxnyxF198URMmTHBpCwoKctn/4IMPFB0drTNnzmjx4sUaNWqUPvnkE8XExDRmqQAAAA2iUR8jBgUFKSwszGVr27atS58OHTooLCxMPXv21EsvvaQLFy4oMzOzzq+5ZMkS9e7dW8HBwQoODlZcXJz+9re/1TgmOztbsbGxat26tW644QYtXbq0zq8PAABatia7ZuvChQt64403JEmtWrWq83k6d+6sl19+Wbm5ucrNzdXPf/5zJSYm6rPPPqu2/7FjxzR8+HANGjRIeXl5mj17tp566imlpaXVuQYAANByNchjxM6dO1dpO3/+fJW2WbNm6fnnn3dp27x5s+68807n/sCBA3XNNdfou+++06VLlxQZGanRo0fXubYRI0a47L/00ktasmSJdu/erejo6Cr9ly5dqi5dujjXm/Xq1Uu5ubmaP3++Ro0aVec6AABAy9QgYWvHjh1V1l79OEBd9uyzz2r8+PEubddff73L/rp169SzZ08dPnxYU6dO1dKlS9W+ffuGKFMXL17UX/7yF5WVlSkuLq7aPrt27VJCQoJL29ChQ5WamqoLFy5Ue5etvLxc5eXlzn2Hw9Eg9QIAgOavQcJWt27d1K5dO9cT+1Y99XXXXafu3bvXeK6IiAj16NFDPXr0UGBgoEaNGqX8/HyFhITUub4DBw4oLi5O33//vQIDA5Wenq6bbrqp2r4lJSUKDQ11aQsNDVVlZaVOnz6t8PDwKmOSk5P1+9//vs71AQAA79Vk12xJ0uDBgxUTE6OXXnqpXueJiorSvn37tHv3bj355JMaN26c8vPzr9jfZrO57FuWVW37ZUlJSSotLXVuJ06cqFe9AADAezTqTz+cPXtWJSUlLm1t2rRRcHDwFcc888wz+vWvf62ZM2dWeeToLj8/P+cdtX79+mnPnj167bXX9Kc//alK37CwsCo1njp1Sr6+vurQoUO15/f395e/v3+dagMAAN6tUe9svfDCCwoPD3fZZs6cWeOYX/7yl4qMjKz33a0fsyzLZY3Vj8XFxSkjI8Olbdu2berXr1+9vhUJAABaJpt1+RmZl5o9e7aGDRumiIgInT17VmvXrtXLL7+srVu36u6771ZSUpKKioq0cuVKST/89ENMTIyeeOIJTZgwQbt27dLEiRO1Zs0at7+N6HA4ZLfbVVpaWuNdOwAA0HSY+vxu1MeInvD1119r7NixKi4ult1uV+/evZ1BS5KKi4tVWFjo7N+tWze9//77mjZtml5//XV16tRJCxcu5GcfAABAnXj9nS1P4M4WAADNj6nP7yb9bUQAAIDmjrAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABjk9WFryZIl6t27t4KDgxUcHKy4uDj97W9/u2L/rKws2Wy2KtuhQ4casWoAAOAtfD1dgGmdO3fWyy+/rO7du0uS3n77bSUmJiovL0/R0dFXHFdQUKDg4GDnfseOHY3XCgAAvI/Xh60RI0a47L/00ktasmSJdu/eXWPYCgkJUbt27QxXBwAAvJ3XP0b8sYsXL2rt2rUqKytTXFxcjX379u2r8PBwxcfHKzMzs8a+5eXlcjgcLhsAAIDUQsLWgQMHFBgYKH9/f02cOFHp6em66aabqu0bHh6uZcuWKS0tTRs2bFBUVJTi4+OVk5NzxfMnJyfLbrc7t4iICFNvBQAANDM2y7IsTxdhWkVFhQoLC3XmzBmlpaXpzTffVHZ29hUD10+NGDFCNptNmzZtqvZ4eXm5ysvLnfsOh0MREREqLS11WfcFAACaLofDIbvd3uCf316/ZkuS/Pz8nAvk+/Xrpz179ui1117Tn/70J7fGDxgwQKtWrbricX9/f/n7+zdIrQAAwLu0iMeIP2VZlsudqKvJy8tTeHi4wYoAAIC38vo7W7Nnz9awYcMUERGhs2fPau3atcrKytLWrVslSUlJSSoqKtLKlSslSSkpKYqMjFR0dLQqKiq0atUqpaWlKS0tzZNvAwAANFNeH7a+/vprjR07VsXFxbLb7erdu7e2bt2qu+++W5JUXFyswsJCZ/+KigrNmDFDRUVFCggIUHR0tLZs2aLhw4d76i0AAIBmrEUskG9sphbYAQAAc0x9frfINVsAAACNhbAFAABgEGELAADAIMIWAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgAAgEGELQAAAIMIWwAAAAYRtgAAAAwibAEAABhE2AIAADCIsAUAAGAQYQsAAMAgwhYAAIBBhC0AAACDCFsAAAAGEbYAAAAM8vV0Ad7IsixJksPh8HAlAADAXZc/ty9/jjcUwpYBZ8+elSRFRER4uBIAAFBb33zzjex2e4Odz2Y1dHyDLl26pJMnTyooKEg2m61e53I4HIqIiNCJEycUHBzcQBXCXcy/5zD3nsX8ew5z7zmlpaXq0qWLvv32W7Vr167BzsudLQOuueYade7cuUHPGRwczP/oPIj59xzm3rOYf89h7j3nmmsadkk7C+QBAAAMImwBAAAYRNhq4vz9/fW73/1O/v7+ni6lRWL+PYe59yzm33OYe88xNfcskAcAADCIO1sAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLDVBCxevFjdunVT69atFRsbqx07dtTYPzs7W7GxsWrdurVuuOEGLV26tJEq9U61mf+srCzZbLYq26FDhxqxYu+Qk5OjESNGqFOnTrLZbNq4ceNVx3DtN4zazj3XfcNJTk5W//79FRQUpJCQEI0cOVIFBQVXHce1X391mfuGuvYJWx62bt06TZ06VXPmzFFeXp4GDRqkYcOGqbCwsNr+x44d0/DhwzVo0CDl5eVp9uzZeuqpp5SWltbIlXuH2s7/ZQUFBSouLnZuPXr0aKSKvUdZWZn69OmjRYsWudWfa7/h1HbuL+O6r7/s7GxNmjRJu3fvVkZGhiorK5WQkKCysrIrjuHabxh1mfvL6n3tW/Con/3sZ9bEiRNd2nr27Gk999xz1fafOXOm1bNnT5e2J554whowYICxGr1Zbec/MzPTkmR9++23jVBdyyHJSk9Pr7EP174Z7sw91705p06dsiRZ2dnZV+zDtW+GO3PfUNc+d7Y8qKKiQnv37lVCQoJLe0JCgj766KNqx+zatatK/6FDhyo3N1cXLlwwVqs3qsv8X9a3b1+Fh4crPj5emZmZJsvE/+Ha9zyu+4ZXWloqSWrfvv0V+3Dtm+HO3F9W32ufsOVBp0+f1sWLFxUaGurSHhoaqpKSkmrHlJSUVNu/srJSp0+fNlarN6rL/IeHh2vZsmVKS0vThg0bFBUVpfj4eOXk5DRGyS0a177ncN2bYVmWpk+frttvv10xMTFX7Me13/DcnfuGuvZ961sw6s9ms7nsW5ZVpe1q/atrh3tqM/9RUVGKiopy7sfFxenEiROaP3++7rjjDqN1gmvfU7juzZg8ebI+/fRT7dy586p9ufYblrtz31DXPne2POi6666Tj49Plbsop06dqvKvmMvCwsKq7e/r66sOHToYq9Ub1WX+qzNgwAAdOXKkocvDT3DtNy1c9/UzZcoUbdq0SZmZmercuXONfbn2G1Zt5r46dbn2CVse5Ofnp9jYWGVkZLi0Z2RkaODAgdWOiYuLq9J/27Zt6tevn1q1amWsVm9Ul/mvTl5ensLDwxu6PPwE137TwnVfN5ZlafLkydqwYYO2b9+ubt26XXUM137DqMvcV6dO1369ltej3tauXWu1atXKSk1NtfLz862pU6dabdu2tY4fP25ZlmU999xz1tixY539v/jiC6tNmzbWtGnTrPz8fCs1NdVq1aqV9e6773rqLTRrtZ3/V1991UpPT7cOHz5s/eMf/7Cee+45S5KVlpbmqbfQbJ09e9bKy8uz8vLyLEnWggULrLy8POvLL7+0LItr36Tazj3XfcN58sknLbvdbmVlZVnFxcXO7fz5884+XPtm1GXuG+raJ2w1Aa+//rrVtWtXy8/Pz7r11ltdvoY6btw4a/DgwS79s7KyrL59+1p+fn5WZGSktWTJkkau2LvUZv5feeUV61/+5V+s1q1bW9dee611++23W1u2bPFA1c3f5a9U/3QbN26cZVlc+ybVdu657htOdfMuyVq+fLmzD9e+GXWZ+4a69m3/VwAAAAAMYM0WAACAQYQtAAAAgwhbAAAABhG2AAAADCJsAQAAGETYAgAAMIiwBQAAYBBhCwAAwCDCFgCPuPPOOzV16lRPl1Gt48ePy2azad++fbUaZ7PZtHHjRiM11da8efN0yy23eLoMACJsAUCz15RCHoCqCFsAmo2KigpPlwAAtUbYAuAxlZWVmjx5stq1a6cOHTro+eef14//XGtkZKT+8Ic/aPz48bLb7ZowYYIkadasWbrxxhvVpk0b3XDDDZo7d64uXLjgHHf5Edqf//xnRUZGym636/7779fZs2edfS5duqRXXnlF3bt3l7+/v7p06aKXXnrJpb4vvvhCQ4YMUZs2bdSnTx/t2rWrVu+vqKhIY8aM0bXXXqsOHTooMTFRx48fdx4fP368Ro4cqfnz5ys8PFwdOnTQpEmTXN5LcXGxfvGLXyggIEDdunXT6tWrFRkZqZSUFOccSdK9994rm83m3L+spjkA0DgIWwA85u2335avr68+/vhjLVy4UK+++qrefPNNlz7/+Z//qZiYGO3du1dz586VJAUFBWnFihXKz8/Xa6+9pjfeeEOvvvqqy7ijR49q48aN2rx5szZv3qzs7Gy9/PLLzuNJSUl65ZVXNHfuXOXn52v16tUKDQ11OcecOXM0Y8YM7du3TzfeeKMeeOABVVZWuvXezp8/ryFDhigwMFA5OTnauXOnAgMDdc8997jcocvMzNTRo0eVmZmpt99+WytWrNCKFSucxx9++GGdPHlSWVlZSktL07Jly3Tq1Cnn8T179kiSli9fruLiYue+O3MAoJFYAOABgwcPtnr16mVdunTJ2TZr1iyrV69ezv2uXbtaI0eOvOq5/uM//sOKjY117v/ud7+z2rRpYzkcDmfbs88+a912222WZVmWw+Gw/P39rTfeeKPa8x07dsySZL355pvOts8++8ySZB08ePCKdUiy0tPTLcuyrNTUVCsqKsrl/ZWXl1sBAQHWf//3f1uWZVnjxo2zunbtalVWVjr7/PrXv7bGjBljWZZlHTx40JJk7dmzx3n8yJEjliTr1VdfrfZ13Z0DAI2HO1sAPGbAgAGy2WzO/bi4OB05ckQXL150tvXr16/KuHfffVe33367wsLCFBgYqLlz56qwsNClT2RkpIKCgpz74eHhzjtCBw8eVHl5ueLj42usr3fv3i7jJbncVarJ3r179fnnnysoKEiBgYEKDAxU+/bt9f333+vo0aPOftHR0fLx8am2zoKCAvn6+urWW291Hu/evbuuvfZat2qoaQ4ANB5fTxcAADVp27aty/7u3bt1//336/e//72GDh0qu92utWvX6o9//KNLv1atWrns22w2Xbp0SZIUEBDg1mv/+ByXQ+Hlc1zNpUuXFBsbq3feeafKsY4dO7pVp/Wj9Ws/dqX2n6rp3AAaD2ELgMfs3r27yn6PHj1c7vT81IcffqiuXbtqzpw5zrYvv/yyVq/bo0cPBQQE6O9//7see+yx2hXtpltvvVXr1q1TSEiIgoOD63SOnj17qrKyUnl5eYqNjZUkff755zpz5oxLv1atWrncDQTQtPAYEYDHnDhxQtOnT1dBQYHWrFmj//qv/9LTTz9d45ju3bursLBQa9eu1dGjR7Vw4UKlp6fX6nVbt26tWbNmaebMmVq5cqWOHj2q3bt3KzU1tT5vx8VvfvMbXXfddUpMTNSOHTt07NgxZWdn6+mnn9ZXX33l1jl69uypu+66S48//rj+53/+R3l5eXr88ccVEBDg8vg1MjJSf//731VSUqJvv/22wd4DgIZB2ALgMQ8//LC+++47/exnP9OkSZM0ZcoUPf744zWOSUxM1LRp0zR58mTdcsst+uijj5zfUqyNuXPn6plnntELL7ygXr16acyYMQ26nqlNmzbKyclRly5d9Ktf/Uq9evXSo48+qu+++65Wd7pWrlyp0NBQ3XHHHbr33ns1YcIEBQUFqXXr1s4+f/zjH5WRkaGIiAj17du3wd4DgIZhs9x9+A8A8LivvvpKERER+uCDD666wB9A00DYAoAmbPv27Tp37pxuvvlmFRcXa+bMmSoqKtLhw4erLIAH0DSxQB4AmrALFy5o9uzZ+uKLLxQUFKSBAwfqnXfeIWgBzQh3tgAAAAxigTwAAIBBhC0AAACDCFsAAAAGEbYAAAAMImwBAAAYRNgCAAAwiLAFAABgEGELAADAoP8HhFgWL5of008AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clade_root = node_to_clade(tree)\n",
    "phylo_tree = Tree(root=clade_root)\n",
    "\n",
    "# Tracer\n",
    "Phylo.draw(phylo_tree, do_show=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "alignment",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
